flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreScanImpl.java

```diff
@@ -109,34 +176,72 @@ public Long snapshotId() {
     }
 
     private List<ManifestEntry> scan() {
-        Map<ManifestEntry.Identifier, ManifestEntry> map = new LinkedHashMap<>();
-        for (ManifestFileMeta manifest : manifests) {
-            // TODO read each manifest file concurrently
-            for (ManifestEntry entry : manifestFile.read(manifest.fileName())) {
-                ManifestEntry.Identifier identifier = entry.identifier();
-                switch (entry.kind()) {
-                    case ADD:
-                        Preconditions.checkState(
-                                !map.containsKey(identifier),
-                                "Trying to add file %s which is already added. "
-                                        + "Manifest might be corrupted.",
-                                identifier);
-                        map.put(identifier, entry);
-                        break;
-                    case DELETE:
-                        Preconditions.checkState(
-                                map.containsKey(identifier),
-                                "Trying to delete file %s which is not previously added. "
-                                        + "Manifest might be corrupted.",
-                                identifier);
-                        map.remove(identifier);
-                        break;
-                    default:
-                        throw new UnsupportedOperationException(
-                                "Unknown value kind " + entry.kind().name());
-                }
+        List<ManifestEntry> entries;
+        try {
+            entries =
+                    new ForkJoinPool()
+                            .submit(
+                                    () ->
+                                            manifests
+                                                    .parallelStream()
+                                                    .filter(this::filterManifestFileMeta)
+                                                    .flatMap(m -> readManifestFileMeta(m).stream())
```

Comment:
filter `filterManifestEntry` in `readManifestFileMeta(m).stream()`?

Reply:
They are the same. Java's stream API is like Flink's job graph. It will not execute immediately but will build a stream graph and will only execute when collected.